---
title: "Intrahost_conspecificity"
output: html_document
date: "2024-10-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Study conspecificity within host

1. To attempt this, we will use the proportion of polymorphic positions found in markers genes used for building the phylogeny

```{r packages}
library(tidyverse)
```

We will define the function to analyse the `*.polymorphic` files generated by StrainPhlAn. 

```{r Function_poly, echo=FALSE}
Process_polymorphic = function( File ){
 df = read.table(File,header = T, sep="\t")
 summary_df <- as.data.frame(as.list(summary(df$percentage_of_polymorphic_sites)))
 return(summary_df)
}

```

Find all `*.polymorphic`, filter out the ones from the SGBs used in study, and run Proces_polymorphic

```{r Prepare data }
read_tsv("/mnt/project/Make_Associations/Association/Tree_sizes2.txt") %>% filter(N_tips >= 300  ) -> SGB_check 
SGB_check$Tree %>% sapply( function(Tree){  str_split(Tree, "\\.")[[1]][2] } ) -> SGBs
SGB_check %>% mutate(SGB_check,SGB = SGBs )  -> SGB_check

polymorphic_files <- list.files(path = "/mnt/project/Phylogenies", full.names = T)
result <- do.call(rbind, lapply(polymorphic_files, function(x) {
  SGB <- basename(x)
  data.frame(SGB = SGB, Polymorphic = paste0(x, "/", SGB, ".polymorphic"), stringsAsFactors = FALSE)
}))
result %>% left_join(SGB_check) %>% as_tibble() %>% drop_na() -> result

```

Per each SGB, get the distribution of conspecificity (% of polymorphic positions) 
```{r iterate }
do.call(rbind, lapply(result$Polymorphic, function(x) {
  SGB <- basename(x) %>% tools::file_path_sans_ext()
  Process_polymorphic(x) %>% mutate(SGB = SGB ) 
})) %>% left_join(result, . , by="SGB") -> result

```

Add taxonomy
```{r tacxonomy}
SGB_taxonomy = read_tsv("/mnt/project/Make_Associations/Phenotypes/mpa_vJan21_CHOCOPhlAnSGB_202103_species.txt.bz2", col_names = F ) %>% rename(SGB=X1, Taxonomy=X2)
SGB_taxonomy <- SGB_taxonomy %>% separate(Taxonomy, into = c("domain","phylum", "class", "order", "family", "genera", "species"), sep = "\\|", remove = FALSE)
SGB_taxonomy %>% mutate(SGB = paste0("t__", SGB  ) ) -> SGB_taxonomy
SGB_taxonomy$species = SGB_taxonomy$species %>% sapply( function(x){ str_split(x, ",")[[1]][1] }  )
SGB_taxonomy %>% select(-Taxonomy) %>% left_join(result, . ) -> result_taxa

```


Check how much conspecificity we actually have, and which lineages have more

```{r compare lineages}
Display = c("SGB", "species", "N_tips",  "Min.", "X1st.Qu.", "Median",  "Mean", "X3rd.Qu.",  "Max.")
result_taxa %>%  select(Display) %>% arrange(desc(Median)) %>% print()
result_taxa %>%  select(Display) %>% arrange(desc(`Max.`)) %>% print()

```

More abundant/prevalent taxa in the population also have greater polymorhic rate within an individual
```{r }
cor.test(result_taxa$Median, result_taxa$N_tips)
```

Let's run an enrichment test at higher levels

```{r }
Run_enrichment = function(Summary_stats, Table,Plot=F){
  Summary_stats %>%  arrange(desc(Median))-> R1
  #Prepare ranks
  ranks = R1$Median
  names(ranks) = as.character(R1$SGB)
  #Run GSEA
  fgseaRes <- fgsea::fgsea(Table, ranks, scoreType = "pos")
  
  if (Plot == T){
    topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
    Plot = fgsea::plotGseaTable(Table[topPathwaysUp], ranks, fgseaRes,  gseaParam=0.5)
  }
  
  return(list('Summary' = fgseaRes, 'Plot' = Plot) )
}
```
Make a list of `[Higher taxonomic level] = c(Species_in_taxa)` which will be used by *fgesea* to run a GSEA analysis
```{r }

Taxa_list = list()
for (i in c("phylum", "class", "order", "family", "genera") ){
  for (Taxa in unique(SGB_taxonomy[[i]]) ){
    SGB_taxonomy %>% filter(!!sym(i) == Taxa ) %>% filter(!is.na(SGB)) -> ForTaxonomy
    Taxa_list[[Taxa]] = ForTaxonomy$SGB
  }
}


```

```{r }
Run_enrichment(result_taxa, Taxa_list, Plot=T)  -> Results_taxa_enrichment  
as_tibble(Results_taxa_enrichment$Summary) %>% arrange(pval)
Results_taxa_enrichment$Plot
```
Check the species in the enriched taxa
```{r }
SGBs_enrichment = arrange(Results_taxa_enrichment$Summary, pval)$leadingEdge[1][[1]]
result_taxa %>% filter(SGB %in% SGBs_enrichment) %>% select( c("genera", Display))

```


