---
title: "RBromii_BMI"
output: html_document
date: "2024-10-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

BMI assocatiom with R. bromii-t__SGB4285_group in European studies.


```{r }
set.seed(12234)
library(tidyverse)
library(anpan)
library(ape)
library(ggtree)
library(ggtreeExtra)
library(ggnewscale)
library(lmerTest)
library(phytools)
library(ggforce)
library(fgsea)

setwd("/mnt/project/Make_Associations/Git/PhylogeneticAssociation/Analyses_revision")
Process_RDS = function(Data){
  read_rds(Data) -> Data
  Data$model_input -> Info
  Data$pglmm_fit$summary() -> Data
  print("Preparing information table")
  Data %>% filter( grepl("phylo", variable))  %>% filter(! grepl("std", variable)) %>% mutate(sample_id = factor(c("-", as.character(Info$sample_id)))) %>%
    select(sample_id, median) %>% rename(phylo_effect_median = median) -> Phylo
  left_join(Info, Phylo) %>% mutate(sample_id = as.character(sample_id)) -> Phylo
  return(Phylo)
}
Process_tree = function(Tree){
  read.tree(Tree) -> Tree
  if (class(Tree) ==  "multiPhylo"){ Tree = Tree[[1]] }
  Tree = midpoint.root(Tree)
  #2. Rename labels tree
  print("Processing tree")
  Tree$tip.label %>% sapply(. , Clean_names) -> New_names
  Tree$tip.label = as.vector(New_names)
  if ( length(Tree$tip.label) < 100 ) { print("Tree is too small") ; q() }
  return(Tree)
  
}
Clean_names =  function(Name){
  'Samples from 500FG FSK need everything after _ to be removed. If _metaphlan4 is attached at the end of the name, remove'
  if ( grepl("HV", Name)) {
    Name = str_split(Name, "_")[[1]][1]
  }   
  Name = str_replace(Name, "_metaphlan4", "") 
  return(Name)
}
Remove_repeated2 = function(Metadata){
  'Check which samples are repeated (either duplicates or longitudinally) and only picks one from the repeated'
  set.seed(89777)
  Metadata %>% group_by(subject_id) %>% summarise(N = n()) %>% arrange(desc(N)) %>% filter(N > 1 ) -> Replicates
  remove = tibble()
  for (i in Replicates$subject_id){
    Metadata %>% filter(subject_id == i) -> Filtered ; Samples = Filtered$ID_anal
    Keep = sample(Samples, 1)
    remove = c(remove, Samples[!Samples==Keep] ) 
  }
  remove = unlist(remove)
  Metadata %>% filter(! ID_anal %in% remove) -> Metadata
  return(Metadata)
}
GetNodeAncesterClade = function(Tree, Bromii_BMI, Threshold=1.40 ){
  #Get the node that descendents in the individuals with higher phylogenetic effect
  if (Threshold > 0){
    Names = filter(Bromii_BMI, phylo_effect_median>Threshold)$sample_id
  } else{
    Names = filter(Bromii_BMI, phylo_effect_median<Threshold)$sample_id
  }
  match( Names , Tree$tip.label) -> CHECK
  common_ancestor <- getMRCA(Tree, CHECK)
  return(common_ancestor)
}
Get_all_descendents = function(MRCA, Tree){
  Tree$tip.label[getDescendants(Tree, MRCA)] -> Clade
  return(Clade[!is.na(Clade)] )
}

```

Prepare data anpan summaries, phylogeny and UniRef90 annotations

```{r pressure, echo=FALSE}
#This take a long time read save and read from a tsv file with the information of interest
#Process_RDS("/mnt/project/Make_Associations/Association/Results/Continent_stratified/Age,Country_Europe/Models/t__SGB4285_group/BMI/Model.rds") -> Bromii_BMI 
#write_tsv(Bromii_BMI, "/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis/BMI_analysis/BMI_Ruminococcus_bromii.tsv")
read_tsv("/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis/BMI_analysis/BMI_Ruminococcus_bromii.tsv") -> Bromii_BMI
Process_tree("/mnt/project/Symlink_phylo/RAxML_bestTree.t__SGB4285_group.TreeShrink.tre") -> Tree_bromii
read_tsv("/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis/Gene_presence/t__SGB4285_group.tsv") -> UniRef90

#Match IDs
UniRef90 %>% filter(Sample %in% Bromii_BMI$sample_id ) -> UniRef90
Bromii_BMI %>% mutate(MAG_available = ifelse(sample_id %in% UniRef90$Sample, T, F) ) -> Bromii_BMI
keep.tip(Tree_bromii,Bromii_BMI$sample_id) -> Tree_bromii

#Stratify BMI
Bromii_BMI %>% mutate(BMI_group = ifelse(BMI<18.5, "BMI<18.5", ifelse(BMI>=18.5 & BMI<25, "18.5≤BMI<25", ifelse(BMI>=25 & BMI<30, "25≤BMI<30", "30≤BMI" ) ) ) ) -> Bromii_BMI


```

Let's see the distribution of phylogenetic effect
```{r}
Bromii_BMI %>% select(phylo_effect_median) %>% arrange(phylo_effect_median) %>% ggplot(aes(x=phylo_effect_median)) + geom_histogram() +theme_bw() + geom_vline(xintercept = 1.4, color="#FF7F7F", linetype=2 ) +  geom_vline(xintercept = -0.6, color="skyblue", linetype=2 )

```

Display tree
```{r }
BMI_palette = c("BMI<18.5" = "#BDF7B7",  "18.5≤BMI<25"= "#F8F4A6", "25≤BMI<30" = "#E08E45", "30≤BMI"= "#6B2737")

Plot_tree = function(Tree_bromii, Bromii_BMI, Clade1, Clade2, Clade3 ){
  
  Node_high = getMRCA(Tree_bromii, (Bromii_BMI %>% filter(sample_id %in% Clade1) )$sample_id  )
  Node_Low = getMRCA(Tree_bromii, (Bromii_BMI %>% filter(sample_id %in% Clade2) )$sample_id  )
  Node_Low2 = getMRCA(Tree_bromii, (Bromii_BMI %>% filter(sample_id %in% Clade3) )$sample_id  )
  
  
  ggtree(Tree_bromii, layout = "fan",  open.angle=15, size=0.1 ) %<+% Bromii_BMI -> p #+
  p + geom_fruit( geom="geom_tile", mapping = aes(fill=BMI_group), width=0.03,offset=0.1  ) + scale_fill_manual(values = BMI_palette, na.value="white" ) +   labs(fill = "BMI") + new_scale_fill() + geom_fruit( geom="geom_tile", mapping = aes(fill=MAG_available), width=0.03,offset=0.1  ) + scale_fill_manual(values = c("FALSE" = "grey",  "TRUE"= "black"), na.value="white"  ) + labs(fill = "MAG") +
new_scale_fill() + geom_fruit( geom="geom_tile", mapping = aes(fill=phylo_effect_median), width=0.03,offset=0.1) + scale_fill_viridis_c(option = "magma") -> Plot
  
  Plot = Plot +  geom_balance(node=Node_high, fill="tomato", color=NA, alpha=0.3) + geom_balance(node=Node_Low, fill="#345995", color=NA, alpha=0.3) + geom_balance(node=Node_Low2  , fill="#03CEA4", color=NA, alpha=0.3) 
    
  return(Plot)  
}
Subsample_tree = function(Tree_bromi, Bromii_BMI, N_samples=1000 ){
  set.seed(12)
  sample(Bromii_BMI$sample_id, N_samples) -> SubSet
  Bromii_BMI %>% filter(sample_id %in% SubSet) -> Bromii_BMI2
  keep.tip(Tree_bromii, SubSet) -> Tree_bromii2
  return(list(Tree = Tree_bromii2, Meta=Bromii_BMI2) )
}

```

The tree is quite large, a subsampling will help visualization
```{r }
Tree_size = Tree_bromii$tip.label %>% length() 
paste0("Number of samples in tree: ", Tree_size)

#Define clades of interest
Node_highBMI = GetNodeAncesterClade(Tree_bromii, Bromii_BMI,Threshold = 1.4)
AllSamplesClade_highBMI = Get_all_descendents(Node_highBMI, Tree_bromii)

Node_lowBMI = GetNodeAncesterClade(Tree_bromii, Bromii_BMI, Threshold = -0.5 )
AllSamplesClade_lowBMI = Get_all_descendents(Node_lowBMI, Tree_bromii)

#Create second 'low BMI clade' outside of the first one (which lacks MAGs)
Node_lowBMI2 = GetNodeAncesterClade(Tree_bromii, Bromii_BMI %>% filter(!sample_id %in% AllSamplesClade_lowBMI), Threshold = -0.35 )
AllSamplesClade_lowBMI2 = Get_all_descendents(Node_lowBMI2, Tree_bromii)


#Subsample tree
Subs = Subsample_tree (Tree_bromii,  Bromii_BMI, 1000)

#Visualize
#Plot_tree( Tree_bromii, Bromii_BMI, Clade1 = AllSamplesClade_highBMI, Clade2 = AllSamplesClade_lowBMI)
Plot_tree( Subs[[1]], Subs[[2]], Clade1 = AllSamplesClade_highBMI, Clade2 = AllSamplesClade_lowBMI, Clade3 = AllSamplesClade_lowBMI2)
```

Check for enrichment of different studies/countries within those clades
```{r }
read_tsv("/mnt/project/Make_Associations/Phenotypes/Phenotypes_merged_complete.tsv") -> Phenos
Phenos %>% select(ID_anal, study_name, Continent, Country, Sex, sample_id) %>% rename(other_id = sample_id ,sample_id = ID_anal) -> Phenos_merge

left_join(Bromii_BMI, Phenos_merge ) -> Bromii_BMI
```

```{r }
Clade_palette = c("Clade_higherBMI"="tomato", "Clade_lowerBMI"="#345995", "Clade_lowerBMI2"="#03CEA4" )

Bromii_BMI %>% mutate( Clade = ifelse(sample_id %in% AllSamplesClade_highBMI, "Clade_higherBMI", 
                               ifelse(sample_id %in% AllSamplesClade_lowBMI, "Clade_lowerBMI",
                               ifelse(sample_id %in% AllSamplesClade_lowBMI2, "Clade_lowerBMI2", "Other") ) ) ) -> Bromii_BMI 
Bromii_BMI %>% mutate(Clade = factor(Clade, levels=c("Other", "Clade_higherBMI","Clade_lowerBMI", "Clade_lowerBMI2" ) )) -> Bromii_BMI

Bromii_BMI %>% ggplot(aes(x=Clade, y=BMI, fill=Clade )) + geom_sina(alpha=0.1) + geom_boxplot(outlier.shape = NA) + theme_bw() + scale_fill_manual(values=Clade_palette)

#Overall association
Bromii_BMI %>% lmer( BMI ~ Clade + Age + Sex + (1|study_name), . ) %>% summary()

```
Country stratification

```{r }
Bromii_BMI %>% group_by(Country) %>% summarise(N = n()) %>% arrange(desc(N)) -> Countries_more_samples

Bromii_BMI %>% filter(Country %in% head(Countries_more_samples,4)$Country ) -> Bromii_BMI_strat

Bromii_BMI_strat %>% ggplot(aes(x=Clade, y=BMI, fill=Clade )) + geom_sina(alpha=0.1) + geom_boxplot(outlier.shape = NA) + theme_bw() + scale_fill_manual(values=Clade_palette) + facet_wrap(~Country) + coord_flip()
```
Stratified association
```{r }
print("Danish association")
Bromii_BMI_strat %>% filter(Country == "DNK") -> Danish_samples
Danish_samples %>% group_by(study_name) %>% summarise(n())
Danish_samples %>% lmer( BMI ~ Clade + Age + Sex + (1|study_name), . ) %>% summary()
```

```{r}
print("Spanish association")
Bromii_BMI_strat %>% filter(Country == "ESP") -> Spanish_samples
Spanish_samples %>% group_by(study_name) %>% summarise(n())
Spanish_samples %>% lmer( BMI ~ Clade + Age + Sex + (1|study_name), . ) %>% summary()
```

```{r}
print("British association")
Bromii_BMI_strat %>% filter(Country == "GBR") -> British_samples
British_samples %>% group_by(study_name) %>% summarise(n())
British_samples %>% lmer( BMI ~ Clade + Age + Sex + (1|study_name), . ) %>% summary()
```

```{r}
print("Dutch association")
Bromii_BMI_strat %>% filter(Country == "NLD") -> Dutch_samples
Dutch_samples %>% group_by(study_name) %>% summarise(n())
Dutch_samples %>% lmer( BMI ~ Clade + Age + Sex + (1|study_name), . ) %>% summary()
```



## Functional enrichment analysis
Check genome quality
```{r }
UniRef90 %>% distinct(Sample, .keep_all = T) %>% ggplot(aes(x=Completeness)) + geom_histogram() + theme_bw()
UniRef90 %>% distinct(Sample, .keep_all = T) %>% ggplot(aes(x=Contamination)) + geom_histogram() + theme_bw()
```

We can do run enrichment at the Gene level or at the product level. Gene level is way bigger.
```{r }
Do_enrichment_uniref90 = function(ToTest=Bromii_BMI ,Feature = "Gene", Subset = T, Uniref = UniRef90, Test ="logistic" ){
  ToTest %>% filter(! is.na(sample_id)) -> ToTest
  if (Subset == T) { 
    Uniref = Uniref %>% filter(Sample %in% filter(ToTest, Clade!="Other")$sample_id ) ->  Uniref
  }
  
  Uniref %>% filter(Sample %in% ToTest$sample_id  )  %>% select( c("Sample", Feature)) %>% distinct(Sample, !!sym(Feature) , .keep_all = T) %>% mutate(Presence = 1) %>% spread( Feature, "Presence") -> UnirefTest
  #All missing values should be 0 for testing
  UnirefTest = UnirefTest %>% distinct(Sample, .keep_all = T)
  UnirefTest[is.na(UnirefTest)] = 0 
  #Add sample information
  UnirefTest %>% left_join( . ,Uniref %>% distinct(Sample, .keep_all =T ) %>% select(Study, Sample, Completeness, Contamination) )  -> UnirefTest
  
  #Keep only one entry per sample
  UnirefTest %>% distinct(Sample, .keep_all =T) -> UnirefTest
  left_join(UnirefTest ,  dplyr::select(ToTest, c(sample_id, Clade)) %>% rename(Sample=sample_id)  ) %>% filter(!is.na(Clade)) -> UnirefTest
  UnirefTest %>% select(-c("Study", "Sample", "Completeness", "Contamination", "Clade")) %>% apply(2,function(x){ mean(x) } ) -> Prevalences
  Results_test = tibble()
  N = 0
  for (Gene in names(Prevalences[Prevalences < 0.9 & Prevalences > 0.1]) ){
    #Add pseudocount
    #if (Gene %in% c("Study", "Sample", "Completeness", "Contamination", "Clade") ){ next }
    mean(UnirefTest[Gene] %>% as_vector() ) -> Prevalence
    #sum( filter(UnirefTest, Clade=="Clade_lowerBMI2"  )[Gene] %>% as_vector()) -> Prevalence_low
    mean( filter(UnirefTest, Clade=="Clade_higherBMI"  )[Gene] %>% as_vector() == 1 ) -> Prevalence_high
    mean( filter(UnirefTest, Clade=="Other"  )[Gene] %>% as_vector() == 1) -> Prevalence_other
    
    N = N + 1    
    #print( c(N ,Gene) )
    
    if (Test == "logistic"){
    
      if (Prevalence_high == 0  ){  
        #tibble(Feature = c("Clade_lowerBMI2", "Clade_higherBMI") , Estimate=NA, `Std. Error`=NA, `z value`=NA,  `Pr(>|z|)`=0, Gene=Gene, Prevalence_high = Prevalence_high, Prevalence_other=Prevalence_other ) %>% rbind(Results_test) ->Results_test 
        next()
       }
        
      Formula = paste0("`",Gene, "` ~ Clade + Completeness + Contamination") 
      UnirefTest %>% glm(Formula , . , family=binomial()) %>% summary() -> Results
      Results$coefficients %>% as.data.frame() %>% rownames_to_column("Feature")  %>% filter(Feature %in%  c("CladeClade_higherBMI", "CladeClade_lowerBMI2" )) %>% as_tibble() %>% mutate(Gene = Gene, Prevalence_high = Prevalence_high, Prevalence_other=Prevalence_other ) %>% rbind(Results_test) -> Results_test
    }
    }
  
  if ( Test == "logistic" ){ Results_test  %>% mutate(FDR = p.adjust(`Pr(>|z|)`, "fdr" ) ) -> Results_test
  } else { Results_test  %>% mutate(FDR = p.adjust(P, "fdr" ) ) -> Results_test }
  print(Results_test)
  
  return(list(Results_test,UnirefTest ))
}  

```


```{r }

Gene_enrichment = Do_enrichment_uniref90(Bromii_BMI , Feature ="Gene", Subset=F,  Uniref = UniRef90, Test ="logistic"  )


```
Check top results
```{r }
Gene_enrichment[[1]] %>% arrange( `Pr(>|z|)` ) %>% select(Gene, Prevalence_high, Prevalence_other, FDR)

Gene_enrichment[[1]] %>% group_by(Estimate>0, FDR<0.05) %>% summarise(n())
```
We see more gene enriched (117) than depleted (4) in this clade with higher BMI. This might be due to the fact that the rest of the tree will have both regions withi/without certain genes, so an enrichment of depleteion is rare.

Finally, we will conduct a pathway enrichment analysis
```{r }
#Metacyc Pathways : reactions
Enrichment_tables = read_rds("/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis/For_enrichment/metacyc_pathways.rds")
#ECs : reactions
Translate_EC_to_reaction = read_tsv("/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis//For_enrichment/EC_to_reaction.tsv.gz")
#Uniref90 to ECs
read_rds("/mnt/project/Make_Associations/Functional_enrichment/MAG_analysis/For_enrichment/UniRef90/ec.rds") -> ECs
ECs_tibble <- enframe(ECs, name = "EC", value = "UniRef90")
ECs_tibble %>% separate_rows(UniRef90, sep="\t") %>% filter(UniRef90 %in% Gene_enrichment[[1]]$Gene ) -> ECs_tibble


Run_enrichment = function(Summary_stats, Table,Plot=F){
  Summary_stats %>%  arrange(desc(`z value`))-> R1
  #Prepare ranks
  ranks = R1$`z value`
  names(ranks) = as.character(R1$Gene)
  #Run GSEA
  fgseaRes <- fgsea::fgsea(Table, ranks)
  
  if (Plot == T){
    topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
    topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
    topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
    fgsea::plotGseaTable(Table[topPathways], ranks, fgseaRes,  gseaParam=0.5)
  }
  
  return(fgseaRes)
}


#1. Uniref90 --> EC ---> Reactions --> Metacyc Enrichment
Gene_enrichment[[1]]  %>% filter(Feature == "CladeClade_higherBMI") %>% mutate(p.adjust(`Pr(>|z|)` , "fdr") ) %>% rename(UniRef90 = Gene) %>% left_join(ECs_tibble) %>% rename(ec = EC) %>% left_join(Translate_EC_to_reaction) %>%  drop_na()  -> Input2

MetaCyc_enrichment = Run_enrichment(Input2 %>% rename(Gene = Reaction) ,  Table=Enrichment_tables )


```

See enrichment results

```{r }
MetaCyc_enrichment %>% as_tibble() %>% arrange(pval)
```
See what was the prevalence of those genes
```{r} 
Significant_pathways = MetaCyc_enrichment %>% as_tibble() %>% filter(padj<0.05)
Significant_pathways$pathway %>% sapply( function(x){ Enrichment_tables[[x]] } ) %>% as_vector() -> Reactions_enriched
Input2 %>% filter(Reaction %in% Reactions_enriched) %>% arrange(Estimate) %>% select(Feature, `z value` , `Pr(>|z|)`, Prevalence_high, Prevalence_other, Pathways)
```
Check enrichment plots
```{r }
r1 = Input2 %>%  arrange(desc(`z value`))
ranks = r1$`z value` ; names(ranks) = as.character(Input2$Reaction)

Make_plot_enr = function(Enrichment_tables, ranks ,Pathway){
  plotEnrichment(Enrichment_tables[[Pathway]], ranks) + labs(title=Pathway) + theme( text = element_text(size = 21),  plot.title = element_text(size = 26), axis.title.x = element_text(size = 21),  axis.title.y = element_text(size = 21)  )
}

Make_plot_enr(Enrichment_tables, ranks ,'PWY-6897')
Make_plot_enr(Enrichment_tables, ranks ,'PWY-7357')
Make_plot_enr(Enrichment_tables, ranks ,'PWY-6527')

```
